core/notification/src/main/java/edu/stanford/spezi/core/notification/fcm/DeviceRegistrationService.kt

```diff
@@ -0,0 +1,108 @@
+package edu.stanford.spezi.core.notification.fcm
+
+import android.content.Context
+import android.os.Build
+import com.google.firebase.functions.FirebaseFunctions
+import dagger.hilt.android.qualifiers.ApplicationContext
+import edu.stanford.spezi.core.coroutines.di.Dispatching
+import edu.stanford.spezi.core.logging.speziLogger
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+import java.util.Locale
+import java.util.TimeZone
+import javax.inject.Inject
+
+/**
+ * Service to register the device with the server.
+ */
+internal class DeviceRegistrationService @Inject constructor(
+    @ApplicationContext private val context: Context,
+    private val functions: FirebaseFunctions,
+    @Dispatching.IO private val ioDispatcher: CoroutineDispatcher,
+) {
+
+    private val logger by speziLogger()
+
+    /**
+     * Registers the device with the server.
+     */
+    fun registerDevice(token: String) {
+        val platform = PLATFORM_ANDROID
+        val osVersion = Build.VERSION.RELEASE
+        val appVersion = getAppVersion()
+        val appBuild = getAppBuild()
+        val language = Locale.getDefault().toLanguageTag()
+        val timeZone = TimeZone.getDefault().id
+
+        val deviceInfo = DeviceInfo(
+            notificationToken = token,
+            platform = platform,
+            osVersion = osVersion,
+            appVersion = appVersion,
+            appBuild = appBuild,
+            language = language,
+            timeZone = timeZone
+        )
+
+        sendDeviceInfoToServer(deviceInfo)
+    }
+
+    private fun getPackageInfo() = context.packageManager.getPackageInfo(context.packageName, 0)
+
+    private fun getAppVersion(): String {
+        return getPackageInfo().versionName
+    }
+
+    private fun getAppBuild(): String {
+        return getPackageInfo().versionCode.toString()
+    }
+
+    private fun sendDeviceInfoToServer(deviceInfo: DeviceInfo) {
+        val deviceInfoMap = with(deviceInfo) {
+            mapOf(
+                KEY_NOTIFICATION_TOKEN to notificationToken,
+                KEY_PLATFORM to platform,
+                KEY_OS_VERSION to osVersion,
+                KEY_APP_VERSION to appVersion,
+                KEY_APP_BUILD to appBuild,
+                KEY_LANGUAGE to language,
+                KEY_TIME_ZONE to timeZone
+            )
+        }
+
+        runCatching {
+            CoroutineScope(ioDispatcher).launch {
```

Comment:
Can we inject `@Dispatching.IO private val coroutineScope: CoroutineScope` directly instead of dispatcher?